File: musify-app/app/src/main/java/com/bombest/musify/data/s3/S3Repository.kt
package com.bombest.musify.data.s3

import android.util.Log
import com.bombest.musify.BuildConfig
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.OkHttpClient
import okhttp3.Request
import java.io.IOException
import java.text.SimpleDateFormat
import java.util.*
import java.util.concurrent.TimeUnit
import java.util.regex.Pattern

/**
 * Repository for fetching tracks from S3 bucket
 * Uses S3 ListObjects API to retrieve track metadata
 */
class S3Repository(
    private val bucket: String = BuildConfig.S3_BUCKET_NAME,
    private val region: String = BuildConfig.S3_REGION,
    private val prefix: String = BuildConfig.S3_PREFIX,
    private val httpClient: OkHttpClient
) {
    companion object {
        private const val TAG = "S3Repository"
        private val AUDIO_EXTENSIONS = setOf(".mp3", ".m4a", ".aac", ".wav", ".flac", ".ogg", ".opus")
        private val ISO_DATE_FORMAT = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", Locale.US).apply {
            timeZone = TimeZone.getTimeZone("UTC")
        }
    }

    /**
     * Fetch all tracks from S3 bucket
     * Parses XML response from S3 ListObjects API
     */
    suspend fun fetchTracks(): Result<List<S3Track>> = withContext(Dispatchers.IO) {
        try {
            Log.i(TAG, "Starting fetchTracks() from bucket=$bucket prefix=$prefix")
            
            val url = "https://$bucket.s3.$region.amazonaws.com/?list-type=2&prefix=$prefix"
            Log.d(TAG, "Fetching from URL: $url")

            val request = Request.Builder()
                .url(url)
                .get()
                .build()

            val response = httpClient.newCall(request).execute()
            
            if (!response.isSuccessful) {
                val errorMessage = when (response.code) {
                    403 -> "Access denied: S3 bucket permissions issue. Please check bucket configuration."
                    404 -> "Bucket not found: S3 bucket does not exist or is not accessible."
                    in 500..599 -> "S3 server error: Please try again later."
                    else -> "Failed to fetch tracks from S3: HTTP ${response.code}"
                }
                Log.e(TAG, errorMessage)
                return@withContext Result.failure(IOException(errorMessage))
            }

            val xmlBody = response.body?.string() ?: ""
            Log.d(TAG, "Received ${xmlBody.length} bytes, parsing XML")

            val tracks = parseTracksFromXml(xmlBody)
            Log.i(TAG, "Successfully parsed ${tracks.size} valid tracks")
            
            tracks.take(3).forEach { track ->
                Log.d(TAG, "  - ${track.extractTrackName()} (${track.size} bytes)")
            }
            if (tracks.size > 3) {
                Log.d(TAG, "  ... and ${tracks.size - 3} more")
            }

            Result.success(tracks)
        } catch (e: IOException) {
            val errorMessage = if (e.message?.contains("timeout", ignoreCase = true) == true) {
                "Network timeout: Unable to fetch tracks from S3. Please check your internet connection."
            } else if (e.message?.contains("socket", ignoreCase = true) == true || 
                      e.message?.contains("host", ignoreCase = true) == true) {
                "Network error: Unable to connect to S3. Please check your internet connection."
            } else {
                "Failed to fetch tracks: ${e.message}"
            }
            Log.e(TAG, errorMessage, e)
            Result.failure(IOException(errorMessage, e))
        } catch (e: Exception) {
            Log.e(TAG, "Unexpected error fetching tracks", e)
            Result.failure(e)
        }
    }

    /**
     * Parse tracks from S3 ListObjects XML response
     */
    private fun parseTracksFromXml(xml: String): List<S3Track> {
        val tracks = mutableListOf<S3Track>()
        var skippedCount = 0

        // Pattern to match <Contents>...</Contents> blocks
        val contentsPattern = Pattern.compile("<Contents>(.*?)</Contents>", Pattern.DOTALL)
        val contentsMatcher = contentsPattern.matcher(xml)

        whileLoop@ while (contentsMatcher.find()) {
            try {
                val contentsBlock = contentsMatcher.group(1) ?: continue

                // Extract Key
                val keyPattern = Pattern.compile("<Key>(.*?)</Key>")
                val keyMatcher = keyPattern.matcher(contentsBlock)
                if (!keyMatcher.find()) {
                    skippedCount++
                    continue
                }
                val key = keyMatcher.group(1)
                if (key == null) {
                    skippedCount++
                    continue@whileLoop
                }

                // Skip "Non-Album" entries (beets placeholder; not a real track)
                if (key.contains("Non-Album", ignoreCase = true)) {
                    skippedCount++
                    continue
                }

                // Skip folders (keys ending with '/')
                if (key.endsWith("/")) {
                    skippedCount++
                    continue
                }

                // Skip entries without extension
                if (!key.contains(".")) {
                    skippedCount++
                    continue
                }

                // Check if it's an audio file
                val lowerKey = key.lowercase()
                val isAudioFile = AUDIO_EXTENSIONS.any { lowerKey.endsWith(it) }
                if (!isAudioFile) {
                    skippedCount++
                    continue
                }

                // Extract Size
                val sizePattern = Pattern.compile("<Size>(.*?)</Size>")
                val sizeMatcher = sizePattern.matcher(contentsBlock)
                val size = if (sizeMatcher.find()) {
                    sizeMatcher.group(1)?.toLongOrNull() ?: 0L
                } else {
                    0L
                }

                // Skip zero-length files
                if (size <= 0) {
                    skippedCount++
                    continue
                }

                // Skip files with numeric-only names (e.g., "00.wav")
                val filename = key.split("/").last()
                if (filename.contains(".")) {
                    val nameWithoutExt = filename.substring(0, filename.lastIndexOf("."))
                    if (nameWithoutExt.matches(Regex("^\\d+$"))) {
                        skippedCount++
                        continue
                    }
                }

                // Extract ETag
                val eTagPattern = Pattern.compile("<ETag>(.*?)</ETag>")
                val eTagMatcher = eTagPattern.matcher(contentsBlock)
                val eTag = if (eTagMatcher.find()) {
                    eTagMatcher.group(1)?.replace("\"", "") ?: ""
                } else {
                    ""
                }

                // Extract LastModified
                val lastModifiedPattern = Pattern.compile("<LastModified>(.*?)</LastModified>")
                val lastModifiedMatcher = lastModifiedPattern.matcher(contentsBlock)
                if (!lastModifiedMatcher.find()) {
                    skippedCount++
                    continue
                }
                val lastModifiedString = lastModifiedMatcher.group(1)
                if (lastModifiedString == null) {
                    skippedCount++
                    continue@whileLoop
                }

                // Parse ISO 8601 date to timestamp
                val lastModified = try {
                    val parsed = ISO_DATE_FORMAT.parse(lastModifiedString)?.time
                    if (parsed == null) {
                        Log.w(TAG, "Invalid date format for $key, skipping")
                        skippedCount++
                        continue@whileLoop
                    }
                    parsed
                } catch (e: Exception) {
                    Log.w(TAG, "Invalid date format for $key, skipping", e)
                    skippedCount++
                    continue@whileLoop
                }

                tracks.add(
                    S3Track(
                        key = key,
                        bucket = bucket,
                        region = region,
                        size = size,
                        lastModified = lastModified,
                        eTag = eTag
                    )
                )
            } catch (e: Exception) {
                // Skip individual track parsing errors, log and continue
                Log.w(TAG, "Error parsing track element, skipping", e)
                skippedCount++
            }
        }

        if (skippedCount > 0) {
            Log.d(TAG, "Skipped $skippedCount invalid entries")
        }

        return tracks
    }

    /**
     * Get track URL for a given S3Track
     */
    fun getTrackUrl(track: S3Track): String = track.url
}

File: spotube-app/lib/plugins/s3_bombest/lib/s3_repository.dart
import 'dart:async';
import 'package:http/http.dart' as http;
import 'package:xml/xml.dart';
import 'models/s3_track.dart';
import 'package:spotube/services/logger/logger.dart';

class S3Repository {
  final String bucket;
  final String region;
  final String prefix;

  S3Repository({
    this.bucket = 'bombest-beats-music',
    this.region = 'us-west-2',
    this.prefix = 'music/',
  });

  String get _baseUrl => "https://$bucket.s3.$region.amazonaws.com";
  static const _audioExtensions = {
    '.mp3',
    '.m4a',
    '.aac',
    '.wav',
    '.flac',
    '.ogg',
    '.opus',
  };

  Future<List<S3Track>> fetchTracks() async {
    AppLogger.log.i('[S3Repository] Starting fetchTracks() from bucket=$bucket prefix=$prefix');
    final url = Uri.parse("$_baseUrl?list-type=2&prefix=$prefix");
    
    AppLogger.log.d('[S3Repository] Fetching from URL: $url');
    
    http.Response response;
    try {
      response = await http.get(url).timeout(
        const Duration(seconds: 30),
        onTimeout: () {
          throw TimeoutException('S3 request timed out after 30 seconds');
        },
      );
    } catch (e, stack) {
      if (e is TimeoutException) {
        AppLogger.log.e('[S3Repository] Request timed out', error: e, stackTrace: stack);
        throw Exception('Network timeout: Unable to fetch tracks from S3. Please check your internet connection.');
      } else if (e.toString().contains('SocketException') || e.toString().contains('Failed host lookup')) {
        AppLogger.log.e('[S3Repository] Network error', error: e, stackTrace: stack);
        throw Exception('Network error: Unable to connect to S3. Please check your internet connection.');
      } else {
        AppLogger.log.e('[S3Repository] Unexpected error', error: e, stackTrace: stack);
        rethrow;
      }
    }

    // Handle different HTTP status codes gracefully
    if (response.statusCode == 200) {
      // Success - continue processing
    } else if (response.statusCode == 403) {
      AppLogger.log.e('[S3Repository] Access forbidden (403) - check bucket permissions');
      throw Exception('Access denied: S3 bucket permissions issue. Please check bucket configuration.');
    } else if (response.statusCode == 404) {
      AppLogger.log.e('[S3Repository] Bucket not found (404)');
      throw Exception('Bucket not found: S3 bucket does not exist or is not accessible.');
    } else if (response.statusCode >= 500) {
      AppLogger.log.e('[S3Repository] Server error ${response.statusCode}');
      throw Exception('S3 server error: Please try again later.');
    } else {
      AppLogger.log.e('[S3Repository] Failed with status ${response.statusCode}');
      throw Exception('Failed to fetch tracks from S3: HTTP ${response.statusCode}');
    }

    AppLogger.log.d('[S3Repository] Received ${response.body.length} bytes, parsing XML');
    
    XmlDocument document;
    try {
      document = XmlDocument.parse(response.body);
    } catch (e, stack) {
      AppLogger.log.e('[S3Repository] XML parsing failed', error: e, stackTrace: stack);
      throw Exception('Failed to parse S3 response: Invalid XML format');
    }
    
    final contents = document.findAllElements('Contents');
    AppLogger.log.d('[S3Repository] Found ${contents.length} Contents elements');

    final tracks = <S3Track>[];
    int skippedCount = 0;
    
    for (final element in contents) {
      try {
        final keyElements = element.findElements('Key');
        if (keyElements.isEmpty) {
          skippedCount++;
          continue;
        }
        final keyElement = keyElements.first;
        
        final key = keyElement.innerText;
        if (key.endsWith('/')) {
          skippedCount++;
          continue; // skip folders
        }
        if (!key.contains('.')) {
          skippedCount++;
          continue; // skip entries without extension
        }
        
        final lowerKey = key.toLowerCase();
        if (_audioExtensions.every((ext) => !lowerKey.endsWith(ext))) {
          skippedCount++;
          continue; // skip non-audio files
        }
        
        final sizeElements = element.findElements('Size');
        final size = sizeElements.isNotEmpty 
            ? int.tryParse(sizeElements.first.innerText) ?? 0 
            : 0;
        if (size <= 0) {
          skippedCount++;
          continue; // skip zero-length
        }
        
        // Skip files with numeric-only names (e.g., "00.wav")
        final filename = key.split('/').last;
        if (filename.contains('.')) {
          final nameWithoutExt = filename.substring(0, filename.lastIndexOf('.'));
          if (RegExp(r'^\d+$').hasMatch(nameWithoutExt)) {
            skippedCount++;
            continue;
          }
        }
        
        final eTagElements = element.findElements('ETag');
        final eTag = eTagElements.isNotEmpty 
            ? eTagElements.first.innerText.replaceAll('"', '')
            : '';
        
        final lastModifiedElements = element.findElements('LastModified');
        if (lastModifiedElements.isEmpty) {
          skippedCount++;
          continue;
        }
        final lastModifiedElement = lastModifiedElements.first;
        
        DateTime lastModified;
        try {
          lastModified = DateTime.parse(lastModifiedElement.innerText);
        } catch (e) {
          AppLogger.log.w('[S3Repository] Invalid date format for $key, skipping');
          skippedCount++;
          continue;
        }

        tracks.add(S3Track(
          key: key,
          bucket: bucket,
          region: region,
          size: size,
          lastModified: lastModified,
          eTag: eTag,
        ));
      } catch (e, stack) {
        // Skip individual track parsing errors, log and continue
        AppLogger.log.w('[S3Repository] Error parsing track element, skipping', error: e, stackTrace: stack);
        skippedCount++;
      }
    }
    
    if (skippedCount > 0) {
      AppLogger.log.d('[S3Repository] Skipped $skippedCount invalid entries');
    }
    
    AppLogger.log.i('[S3Repository] Successfully parsed ${tracks.length} valid tracks');
    for (final track in tracks.take(3)) {
      AppLogger.log.d('[S3Repository]   - ${track.title} (${track.size} bytes)');
    }
    if (tracks.length > 3) {
      AppLogger.log.d('[S3Repository]   ... and ${tracks.length - 3} more');
    }
    
    return tracks;
  }

  String getTrackUrl(S3Track track) {
    return track.url;
  }
}
